odoo/tools/commit.py

```diff
@@ -0,0 +1,301 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+"""Transaction commit-related utilities.
+
+In general, in Odoo code commit() should never be called.
+There are some exceptions, like ir.cron runs where you want to commit during
+the run. It may be error prone, this module standardises this.
+In other cases, you may use `cr.savepoint()` instead.
+"""
+
+import datetime
+import logging
+import threading
+
+from odoo.exceptions import ValidationError
+from odoo.models import Model
+from odoo.osv import expression
+
+from .misc import split_every
+
+LOG_FREQUENCY = datetime.timedelta(seconds=5)
+MAX_SEARCH_RECORDS = 1000
+
+
+def auto_commit(env):
+    """Returns either commit() or flush_all() depending on the context"""
+    should_commit = bool(
+        env.context.get('auto_commit', True)
```

Comment:
The main goal is to make multi-transaction methods forbidden and offer a way in the framework to manage multiple related transactions. After a commit, all cache must be invalidated.
The way I'm using this code: create a model to handle the work (example: mail output); users will write a records for work to be done and a cron can be triggered to handle them (one by one). The concept can be generalized to things to do after the transaction is successful.

Committing after each SQL has no place in an ORM, it's usually quite dangerous and used only interactively (as far as I know).
I don't think `auto_commit` should be acceptable. For example I had a case where we wanted to create and send multiple documents (invoices); and there was an excption for one of the documents - that ended up with a few inconsistencies and parially rolled-back transactions (because I was not aware that the function I used for sending an e-mail did auto-commit).

In general, PDF generation, mail sending and similar action could be done outside of the main transaction. The method can register a small task to do after the commit (and could even join/wait for completion of the task in the postcommit before returning the HTTP call result).

Anyways, I'm starting working at Odoo next week so we can discuss it live.